File: ccp/app/src/lib/dependsOn.ts
import type { ManifestParam } from '@/types';

/**
 * The ONE conditional-visibility predicate.
 *
 * A param with `dependsOn` is ACTIVE only while its controlling param's
 * current value satisfies the condition; while inactive it is not rendered,
 * never required, its bounds are not applied, and its value is stripped from
 * the submitted request. This module is imported by BOTH the SPA
 * (SchemaForm render + interpreter.validateParams + buildRequestDraft) and
 * ccp-api (`@app-lib/dependsOn` in src/manifests.ts validateParams), so
 * the client can never show a rule the server does not enforce, and the
 * server never demands a field the form deliberately hid — one predicate,
 * zero drift. Pure and deterministic: string-canonical comparison
 * over the submitted values, no I/O.
 *
 * Semantics (documented on types/manifest.ts ParamCondition):
 *   - `equals`:    active ⇔ String(controller) === String(equals)
 *   - `notEquals`: active ⇔ controller is NON-EMPTY and differs — an unset
 *     controller deactivates the dependent under both operators (fail-closed
 *     to hidden; a blank protocol never reveals port fields).
 *   - no dependsOn ⇒ always active (pre-existing behavior exactly).
 */

function isEmpty(v: unknown): boolean {
  return v === undefined || v === null || v === '';
}

/** True while `param` should render and validate, given the current values. */
export function isParamActive(
  param: Pick<ManifestParam, 'dependsOn'>,
  values: Record<string, unknown>,
): boolean {
  const cond = param.dependsOn;
  if (!cond) return true;
  const controller = values[cond.param];
  if (isEmpty(controller)) return false;
  if (cond.equals !== undefined) return String(controller) === String(cond.equals);
  if (cond.notEquals !== undefined) return String(controller) !== String(cond.notEquals);
  // A condition with neither operator cannot validate (zod refuses it at
  // load); treat it as never-active so a malformed manifest fails closed.
  return false;
}

/**
 * The submit-side twin of the render gate: drop every value whose param is
 * currently inactive, so hidden-field values (stale defaults, values left
 * behind by a controller change) never reach the stored request or the pinned
 * artifact. Keys that do not correspond to a declared param pass through
 * unchanged (pre-existing behavior for the target address and shell fields).
 */
export function stripInactiveParams(
  params: Pick<ManifestParam, 'name' | 'dependsOn'>[],
  values: Record<string, unknown>,
): Record<string, unknown> {
  const inactive = new Set(
    params.filter((p) => !isParamActive(p, values)).map((p) => p.name),
  );
  if (inactive.size === 0) return values;
  const out: Record<string, unknown> = {};
  for (const [k, v] of Object.entries(values)) {
    if (!inactive.has(k)) out[k] = v;
  }
  return out;
}

File: tools/catalogctl/internal/manifests/dependson.go
package manifests

// dependson.go is the Go port of app/src/lib/dependsOn.ts#isParamActive — the ONE
// conditional-visibility predicate. The TS module is imported by BOTH
// the SPA and ccp-api's validateParams; this is its executor twin so the Go
// bounds re-check (Validate) and the create renderer gate identically to the form —
// "one predicate, zero drift". A param with dependsOn is ACTIVE only while its
// controller's value satisfies the condition; while inactive it is not rendered,
// NEVER required, and its bounds are not applied (Validate skips it entirely). Pure
// and deterministic.

import (
	"encoding/json"
	"fmt"
)

// paramCondition mirrors the ParamCondition zod shape: a controlling param plus
// exactly one of equals/notEquals. RawMessage distinguishes a present operator from
// an absent one (the TS `cond.equals !== undefined` gate).
type paramCondition struct {
	Param     string          `json:"param"`
	Equals    json.RawMessage `json:"equals"`
	NotEquals json.RawMessage `json:"notEquals"`
}

// IsParamActive reports whether p should render and validate given the current
// values. No dependsOn ⇒ always active (pre-existing behavior exactly). An unset or
// empty controller deactivates the dependent under both operators (fail-closed to
// hidden). A malformed condition also fails closed.
func IsParamActive(p Param, values map[string]any) bool {
	if !rawPresent(p.DependsOn) {
		return true
	}
	var cond paramCondition
	if err := json.Unmarshal(p.DependsOn, &cond); err != nil {
		return false
	}
	controller, ok := values[cond.Param]
	if !ok || isEmptyController(controller) {
		return false
	}
	if rawPresent(cond.Equals) {
		return fmt.Sprint(controller) == fmt.Sprint(rawToAny(cond.Equals))
	}
	if rawPresent(cond.NotEquals) {
		return fmt.Sprint(controller) != fmt.Sprint(rawToAny(cond.NotEquals))
	}
	return false
}

// isEmptyController mirrors dependsOn.ts#isEmpty (undefined/null/"" only).
func isEmptyController(v any) bool {
	switch x := v.(type) {
	case nil:
		return true
	case string:
		return x == ""
	}
	return false
}

// rawPresent reports whether a json.RawMessage carries a real (non-null) value.
func rawPresent(raw json.RawMessage) bool {
	s := string(raw)
	return len(raw) > 0 && s != "null"
}

// rawToAny unmarshals a json.RawMessage into a Go any, or nil on failure.
func rawToAny(raw json.RawMessage) any {
	var v any
	if err := json.Unmarshal(raw, &v); err != nil {
		return nil
	}
	return v
}
